File: frontend/src/lib/api/studio/audio.ts
/**
 * Audio Overview API
 * Educational Note: Handles audio overview generation using ElevenLabs TTS.
 * Non-blocking pattern - returns job_id for polling.
 */

import axios from 'axios';
import { API_BASE_URL } from '../client';
import type { JobStatus } from './index';
import { createLogger } from '@/lib/logger';

const log = createLogger('studio-audio-api');

/**
 * Audio job status (alias for backwards compatibility)
 */
export type AudioJobStatus = JobStatus;

/**
 * Audio job record from the API
 */
export interface AudioJob {
  id: string;
  source_id: string;
  source_name: string;
  direction: string;
  status: AudioJobStatus;
  progress: string;
  error: string | null;
  audio_path: string | null;
  audio_filename: string | null;
  audio_url: string | null;
  script_path: string | null;
  audio_info: {
    file_size_bytes?: number;
    estimated_duration_seconds?: number;
    word_count?: number;
    voice_id?: string;
    model_id?: string;
  } | null;
  // Edit lineage
  parent_job_id: string | null;
  edit_instructions: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

/**
 * Response from starting audio generation
 */
export interface StartAudioResponse {
  success: boolean;
  job_id?: string;
  message?: string;
  source_name?: string;
  error?: string;
}

/**
 * Response from getting job status
 */
export interface AudioJobStatusResponse {
  success: boolean;
  job?: AudioJob;
  error?: string;
}

/**
 * Response from listing jobs
 */
export interface ListAudioJobsResponse {
  success: boolean;
  jobs: AudioJob[];
  count: number;
  error?: string;
}

/**
 * TTS configuration status
 */
export interface TTSStatusResponse {
  success: boolean;
  configured: boolean;
  error?: string;
}

/**
 * Audio Overview API
 */
export const audioAPI = {
  /**
   * Start audio overview generation or edit
   * Educational Note: Non-blocking - returns immediately with job_id
   */
  async startGeneration(
    projectId: string,
    sourceId: string,
    direction?: string,
    parentJobId?: string,
    editInstructions?: string
  ): Promise<StartAudioResponse> {
    try {
      const body: Record<string, unknown> = {
        source_id: sourceId,
        direction: direction || 'Create an engaging audio overview of this content.',
      };
      if (parentJobId) body.parent_job_id = parentJobId;
      if (editInstructions) body.edit_instructions = editInstructions;

      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/studio/audio-overview`,
        body
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to start audio generation');
      throw error;
    }
  },

  /**
   * Get the status of an audio generation job
   */
  async getJobStatus(projectId: string, jobId: string): Promise<AudioJobStatusResponse> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to get job status');
      throw error;
    }
  },

  /**
   * List all audio jobs for a project
   */
  async listJobs(projectId: string, sourceId?: string): Promise<ListAudioJobsResponse> {
    try {
      const params = sourceId ? { source_id: sourceId } : {};
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/jobs`,
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to list jobs');
      throw error;
    }
  },

  /**
   * Check if TTS (ElevenLabs) is configured
   */
  async checkTTSStatus(): Promise<TTSStatusResponse> {
    try {
      const response = await axios.get(`${API_BASE_URL}/studio/tts/status`);
      return response.data;
    } catch (error) {
      log.error({ err: error }, 'failed to check TTS status');
      return { success: false, configured: false };
    }
  },

  /**
   * Get the full URL for an audio file
   */
  getAudioUrl(projectId: string, jobId: string, filename: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/audio/${jobId}/${filename}`;
  },

  /**
   * Poll job status until complete or error
   * Educational Note: Uses polling with exponential backoff
   */
  async pollJobStatus(
    projectId: string,
    jobId: string,
    onProgress?: (job: AudioJob) => void,
    intervalMs: number = 2000,
    maxAttempts: number = 120
  ): Promise<AudioJob> {
    let attempts = 0;
    let currentInterval = intervalMs;

    while (attempts < maxAttempts) {
      const response = await this.getJobStatus(projectId, jobId);

      if (!response.success || !response.job) {
        throw new Error(response.error || 'Failed to get job status');
      }

      const job = response.job;

      if (onProgress) {
        onProgress(job);
      }

      if (job.status === 'ready' || job.status === 'error') {
        return job;
      }

      await new Promise((resolve) => setTimeout(resolve, currentInterval));

      attempts++;

      if (attempts > 5 && currentInterval < 5000) {
        currentInterval = Math.min(currentInterval * 1.2, 5000);
      }
    }

    throw new Error('Audio generation timed out');
  },

  /**
   * Delete an audio job
   */
  async deleteJob(projectId: string, jobId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const response = await axios.delete(
        `${API_BASE_URL}/projects/${projectId}/studio/audio-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to delete audio job');
      throw error;
    }
  },
};

File: backend/app/api/studio/audio.py
"""
Audio Overview endpoints - TTS-based audio summaries.

Educational Note: Audio overviews demonstrate text-to-speech (TTS) integration:
1. AI generates a conversational script from source content
2. ElevenLabs API converts script to natural-sounding audio
3. Audio file is stored and served for playback

TTS Integration Pattern:
- ElevenLabs provides high-quality voice synthesis
- Voices can be selected from their library (professional narrators)
- Audio is generated in MP3 format for web playback

Background Job Flow:
1. POST /audio-overview creates job, returns job_id
2. audio_overview_service runs in background thread
3. Frontend polls GET /jobs/{job_id} for status
4. When ready, audio_url points to GET /audio/<filename>

Routes:
- POST   /projects/<id>/studio/audio-overview       - Start generation
- GET    /projects/<id>/studio/jobs/<id>            - Job status
- GET    /projects/<id>/studio/jobs                 - List jobs
- GET    /projects/<id>/studio/audio/<job_id>/<file> - Serve audio file
- DELETE /projects/<id>/studio/audio-jobs/<id>      - Delete job
- GET    /studio/tts/status                         - Check TTS config
- GET    /studio/tts/voices                         - List available voices
"""
import io
import uuid
from flask import jsonify, request, current_app, send_file
from app.api.studio import studio_bp
from app.services.studio_services import audio_overview_service, studio_index_service
from app.services.source_services import source_index_service
from app.services.integrations.elevenlabs import tts_service
from app.services.integrations.supabase import storage_service  # For downloading previous scripts during edits
from app.services.background_services.task_service import task_service
from app.services.auth import require_permission


@studio_bp.route('/projects/<project_id>/studio/audio-overview', methods=['POST'])
@require_permission("studio", "audio_overview")
def generate_audio_overview(project_id: str):
    """
    Start audio overview generation or edit as a background task.

    Educational Note: This endpoint is non-blocking:
    1. Creates a job record with status="pending"
    2. Submits background task via task_service
    3. Returns job_id immediately for status polling

    For edits, the previous script is loaded and passed to the agent as context,
    so Claude can refine the audio based on user instructions.

    Request Body:
        - source_id: UUID of the source to generate overview for (required)
        - direction: Optional guidance for the script style/focus
        - parent_job_id: UUID of the parent job to edit (optional, for edits)
        - edit_instructions: Instructions for editing the parent audio (optional, for edits)

    Response:
        - success: Boolean
        - job_id: ID for polling status
        - message: Status message
    """
    try:
        data = request.get_json() or {}

        source_id = data.get('source_id')

        direction = data.get('direction', 'Create an engaging audio overview of this content.')

        # Check if TTS is configured
        if not tts_service.is_configured():
            return jsonify({
                'success': False,
                'error': 'ElevenLabs API key not configured. Please add it in Admin Settings.'
            }), 400

        # Edit mode: load parent job's script as context for refinement
        parent_job_id = data.get('parent_job_id')
        edit_instructions = data.get('edit_instructions')
        previous_content = None

        if parent_job_id and not edit_instructions:
            return jsonify({
                'success': False,
                'error': 'edit_instructions is required when parent_job_id is provided'
            }), 400

        if parent_job_id:
            # Clean up any previously failed edit jobs for this parent
            try:
                all_jobs = studio_index_service.list_audio_jobs(project_id)
                for job in all_jobs:
                    if (job.get("status") == "error"
                            and job.get("parent_job_id") == parent_job_id):
                        studio_index_service.delete_audio_job(project_id, job["id"])
            except Exception:
                pass  # Non-critical cleanup -- don't block the new edit

            parent_job = studio_index_service.get_audio_job(project_id, parent_job_id)
            if not parent_job:
                return jsonify({
                    'success': False,
                    'error': 'Parent job not found'
                }), 404

            # Download the previous script from Supabase Storage
            previous_content = storage_service.download_studio_file(
                project_id=project_id,
                job_type="audio",
                job_id=parent_job_id,
                filename="script.txt"
            )
            if previous_content is None:
                return jsonify({
                    'success': False,
                    'error': 'Could not load previous audio script for editing. Please try regenerating.'
                }), 422

            # Inherit source_id from parent if not explicitly provided
            if not source_id:
                source_id = parent_job.get('source_id')

        # Now enforce source_id after potential inheritance
        if not source_id:
            return jsonify({
                'success': False,
                'error': 'source_id is required'
            }), 400

        # Get source info for the job record
        source = source_index_service.get_source_from_index(project_id, source_id)
        if not source:
            return jsonify({
                'success': False,
                'error': f'Source not found: {source_id}'
            }), 404

        source_name = source.get('name', 'Unknown')

        # Create job record
        job_id = str(uuid.uuid4())
        studio_index_service.create_audio_job(
            project_id=project_id,
            job_id=job_id,
            source_id=source_id,
            source_name=source_name,
            direction=direction,
            parent_job_id=parent_job_id,
            edit_instructions=edit_instructions
        )

        # Submit background task
        task_service.submit_task(
            task_type="audio_overview",
            target_id=job_id,
            callable_func=audio_overview_service.generate_audio_overview,
            project_id=project_id,
            source_id=source_id,
            job_id=job_id,
            direction=direction,
            previous_content=previous_content,
            edit_instructions=edit_instructions
        )

        return jsonify({
            'success': True,
            'job_id': job_id,
            'message': 'Audio generation started',
            'source_name': source_name
        }), 202  # 202 Accepted - processing started

    except Exception as e:
        current_app.logger.error(f"Error starting audio overview: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to start audio generation: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/jobs/<job_id>', methods=['GET'])
def get_audio_job_status(project_id: str, job_id: str):
    """
    Get the status of an audio generation job.

    Educational Note: Frontend polls this endpoint to check progress.
    When status="ready", the audio_url field contains the playback URL.

    Response:
        - success: Boolean
        - job: Job record with status, progress, audio_url (when ready)
    """
    try:
        job = studio_index_service.get_audio_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Job not found: {job_id}'
            }), 404

        return jsonify({
            'success': True,
            'job': job
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error getting job status: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to get job status: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/jobs', methods=['GET'])
def list_audio_jobs(project_id: str):
    """
    List all audio generation jobs for a project.

    Query Parameters:
        - source_id: Optional filter by source

    Response:
        - success: Boolean
        - jobs: List of job records
    """
    try:
        source_id = request.args.get('source_id')
        jobs = studio_index_service.list_audio_jobs(project_id, source_id)

        # Filter out orphaned failed-edit jobs (error + parent_job_id).
        clean_jobs = [
            job for job in jobs
            if not (job.get("status") == "error" and job.get("parent_job_id"))
        ]

        return jsonify({
            'success': True,
            'jobs': clean_jobs,
            'count': len(clean_jobs)
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error listing jobs: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to list jobs: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/audio/<job_id>/<filename>', methods=['GET'])
def get_audio_file(project_id: str, job_id: str, filename: str):
    """
    Serve an audio file from Supabase Storage.

    Educational Note: This endpoint serves generated audio files for playback.
    The frontend can use this URL as the src for an <audio> element.

    Response:
        - Audio file (mp3) with appropriate headers for streaming
    """
    try:
        data = storage_service.download_studio_binary(
            project_id, "audio", job_id, filename
        )

        if not data:
            return jsonify({
                'success': False,
                'error': f'Audio file not found: {filename}'
            }), 404

        return send_file(io.BytesIO(data), mimetype='audio/mpeg', as_attachment=False)

    except Exception as e:
        current_app.logger.error(f"Error serving audio file: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to serve audio file: {str(e)}'
        }), 500


@studio_bp.route('/studio/tts/status', methods=['GET'])
def get_tts_status():
    """
    Check if TTS (ElevenLabs) is configured.

    Educational Note: This endpoint allows the frontend to check
    if audio generation is available before showing the option.

    Response:
        - configured: Boolean indicating if ElevenLabs API key is set
    """
    try:
        return jsonify({
            'success': True,
            'configured': tts_service.is_configured()
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error checking TTS status: {e}")
        return jsonify({
            'success': False,
            'error': 'Failed to check TTS status'
        }), 500


@studio_bp.route('/studio/tts/voices', methods=['GET'])
def list_tts_voices():
    """
    List available TTS voices from ElevenLabs.

    Educational Note: This endpoint allows users to choose
    their preferred voice for audio overviews.

    Response:
        - success: Boolean
        - voices: List of voice info (id, name, category, preview_url)
    """
    try:
        if not tts_service.is_configured():
            return jsonify({
                'success': False,
                'error': 'ElevenLabs API key not configured'
            }), 400

        result = tts_service.list_voices()
        return jsonify(result), 200 if result.get('success') else 400

    except Exception as e:
        current_app.logger.error(f"Error listing TTS voices: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to list voices: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/audio-jobs/<job_id>', methods=['DELETE'])
def delete_audio_job(project_id: str, job_id: str):
    """
    Delete an audio job and its files.

    Response:
        - Success status
    """
    try:
        job = studio_index_service.get_audio_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Audio job {job_id} not found'
            }), 404

        # Delete files from Supabase Storage
        storage_service.delete_studio_job_files(project_id, "audio", job_id)

        # Delete job from index
        studio_index_service.delete_audio_job(project_id, job_id)

        return jsonify({
            'success': True,
            'message': f'Audio job {job_id} deleted'
        })

    except Exception as e:
        current_app.logger.error(f"Error deleting audio job: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to delete audio job: {str(e)}'
        }), 500
